### PR TITLE
Don't transfer objects whose data is not needed

### PR DESCRIPTION
**Pull request by stephanie-wang**
## What do these changes do?

The task dependency manager makes any needed object dependencies local, by pulling from a remote object manager if necessary. However, tasks that call `ray.wait` do not actually need the data transferred. Also, actor dummy objects have no data and therefore can never be pulled from a remote object manager. This PR prevents transfer of these objects by adding an extra `request_transfer` flag to the task dependency manager's methods.

### CODE REVIEW

**Review by ericl — COMMENTED** [review]
What is the current behavior on master, would it not return until the transfer completed? I suspect this is actually desirable for most use cases. Otherwise it is impossible to write a non blocking event loop using Ray.wait

Maybe add a flag to allow not waiting for the data fetch?

File: src/ray/raylet/task_dependency_manager.h [hunk 1]
```diff
@@ -143,7 +153,7 @@ class TaskDependencyManager {
   /// transfer or reconstruction. These are objects for which: (1) there is a
   /// subscribed task dependent on it, (2) the object is not local, and (3) the
   /// task that creates the object is not pending execution locally.
```
Comment (robertnishihara) [inline src/ray/raylet/task_dependency_manager.h]: Could you document the arguments here? Especially if `transfer_required` is an output.

**Comment by robertnishihara** [comment]
@stephanie-wang @ericl what is the motivating use case for this PR? In many cases, `ray.wait` is followed immediately by `ray.get`, and the items that were not ready are then processed in a subsequent call to `ray.wait`. So requesting the transfer can improve performance.

I suppose it could also lead to excess transfers depending on how it is used. Have we been running into problems with this?

**Comment by stephanie-wang** [comment]
Temporarily closing so we can make some progress on higher priority PRs.

Note from offline discussion: we will likely add a flag to `ray.wait` to indicate whether object transfer is needed.

**Pull request closed without merging**